pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,599 @@
+                            log.error("[{}] Failed to open readOnly managed ledger", topic, exception);
+                            openManagedLedgerFuture.completeExceptionally(exception);
+                        }
+                    };
+
+                    TopicName snapshotIndexTopicName = TopicName.get(TopicDomain.persistent.toString(),
+                            TopicName.get(topic.getName()).getNamespaceObject(),
+                            EventType.TRANSACTION_BUFFER_SNAPSHOT_SEGMENTS.toString());
+                    this.topic.getBrokerService().getPulsar().getManagedLedgerFactory()
+                            .asyncOpenReadOnlyManagedLedger(snapshotIndexTopicName
+                                            .getPersistenceNamingEncoding(), callback,
+                                    topic.getManagedLedger().getConfig(),
+                                    null);
+                    //Wait the processor recover completely and the allow TB to recover the messages
+                    // after the startReadCursorPosition.
+
+                    return openManagedLedgerFuture
+                            .thenCompose((ignore) -> FutureUtil.waitForAll(completableFutures).thenCompose((i) -> {
+                                if (invalidIndex.get() != 0) {
+                                    persistentWorker.appendTask(PersistentWorker.OperationType.UpdateIndex, ()
+                                            -> persistentWorker
+                                            .updateSnapshotIndex(persistentSnapshotIndexes.getSnapshot(),
+                                            indexes.values().stream().toList()));
+                                }
+                                return CompletableFuture.completedFuture(finalStartReadCursorPosition);
+                            })).exceptionally(ex -> {
+                                log.error("[{}] Failed to recover snapshot segment", this.topic.getName(), ex);
+                                return null;
+                            });
+
+                    },  topic.getBrokerService().getPulsar().getTransactionExecutorProvider()
+                        .getExecutor(this));
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteAbortedTxnSnapshot() {
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        persistentWorker.appendTask(PersistentWorker.OperationType.Close,
+                () -> persistentWorker.clearSnapshotSegmentAndIndexes()
+                        .thenRun(() -> {
+                            completableFuture.thenCompose(null);
+                        }).exceptionally(e -> {
+                            completableFuture.completeExceptionally(e);
+                            return null;
+                        }));
+        return completableFuture;
+    }
+
+    @Override
+    public long getLastSnapshotTimestamps() {
+        return this.lastSnapshotTimestamps;
```

Comment:
`lastSnapshotTimestamps` was also changed when doing `recover` or `delete segment`. So what does this variable really mean?

Reply:
It can be understood to only change in `updateSnapshotIndex`.
When the latest max read position and aborted txn id are persistent in the index snapshot, lastSnapshotTimestamps will be updated.